src/main/java/org/folio/circulation/services/LostItemFeeRefundService.java

```diff
@@ -0,0 +1,136 @@
+      checkInRecords.getItemStatusBeforeCheckIn(), checkInRecords.getLoan(),
+      checkInRecords.getLoggedInUserId(), checkInRecords.getCheckInServicePointId().toString());
+
+    return refundLostItemFees(referenceDataContext)
+      .thenApply(r -> r.map(notUsed -> checkInRecords));
+  }
+
+  private CompletableFuture<Result<Void>> refundLostItemFees(ReferenceDataContext context) {
+    if (context.itemStatus != ItemStatus.DECLARED_LOST) {
+      return completedFuture(succeeded(null));
+    }
+
+    return fetchLostItemPolicy(succeeded(context))
+      .thenCompose(contextResult -> contextResult.after(refData -> {
+        if (!refData.lostItemPolicy.shouldRefundFees(refData.loan.getDeclareLostDateTime())) {
+          log.debug("Refund interval has exceeded for loan [{}]", refData.loan.getId());
+          return completedFuture(succeeded(null));
+        }
+
+        return fetchAccountsAndActionsForLoan(contextResult)
+          .thenCompose(r -> r.after(notUsed -> feeFineService
+            .refundAndCloseAccounts(getAccountsToRefund(context))))
+          .thenApply(r -> r.map(notUsed -> null));
+      }));
+  }
+
+  private List<RefundAccountCommand> getAccountsToRefund(ReferenceDataContext context) {
+    Collection<Account> allAccounts = context.accounts;
+
+    if (!context.lostItemPolicy.isRefundProcessingFeeWhenReturned()) {
+      allAccounts = context.accounts.stream()
+        .filter(account -> !account.getFeeFineType().equals(LOST_ITEM_PROCESSING_FEE_TYPE))
+        .collect(Collectors.toList());
+    }
+
+    return allAccounts.stream()
+      .map(account -> new RefundAccountCommand(account, context.staffUserId, context.servicePointId))
+      .collect(Collectors.toList());
+  }
+
+  private CompletableFuture<Result<ReferenceDataContext>> fetchAccountsAndActionsForLoan(
+    Result<ReferenceDataContext> contextResult) {
+
+    return contextResult.after(context -> {
+      final Result<CqlQuery> fetchQuery = exactMatch("loanId", context.loan.getId())
+        .combine(exactMatchAny("feeFineType", LOST_ITEM_FEE_TYPES), CqlQuery::and);
+
+      return accountRepository.findAccountsAndActions(fetchQuery)
+        .thenApply(r -> r.map(context::withAccounts));
+    });
+  }
+
+  private CompletableFuture<Result<ReferenceDataContext>> fetchLostItemPolicy(
+    Result<ReferenceDataContext> contextResult) {
+
+    return contextResult.combineAfter(
+      context -> lostItemPolicyRepository.getLostItemPolicyById(context.loan.getLostItemPolicyId()),
+      ReferenceDataContext::withLostItemPolicy);
+  }
+
+  private static final class ReferenceDataContext {
+    private final ItemStatus itemStatus;
+    private final Loan loan;
+    private final String staffUserId;
+    private final String servicePointId;
+    private Collection<Account> accounts;
+    private LostItemPolicy lostItemPolicy;
+
+    public ReferenceDataContext(ItemStatus itemStatus, Loan loan, String staffUserId,
```

Comment:
Can this be private, is it used outside of the class?

Reply:
Sure, it is private now.